dali/operators/reader/loader/webdataset_loader.cc

```diff
@@ -0,0 +1,437 @@
+      sample_was_set[component_index] = true;
+    }
+  }
+}
+
+
+inline uint8_t* WebdatasetLoader::ShareDataPointer(std::vector<Tensor<CPUBackend>>& sample,
+                                                   std::vector<char>& sample_was_set,
+                                                   const std::string& extension,
+                                                   const std::string& source_info,
+                                                   int64_t size) const {
+  DALIMeta meta;
+  meta.SetSourceInfo(source_info);
+  meta.SetSkipSample(false);
+
+  uint8_t* shared_tensor_data = nullptr;
+  for (size_t component_index : ext_map_.at(extension)) {
+    if (!sample_was_set[component_index]) {
+      sample[component_index].SetMeta(meta);
+      auto dtype_info = TypeTable::GetTypeInfo(dtypes_[component_index]);
+      if (shared_tensor_data == nullptr) {
+        if (sample[component_index].shares_data()) {
+          sample[component_index].Reset();
+        }
+        sample[component_index].reserve(size);
+        sample[component_index].Resize({size / static_cast<int64_t>(dtype_info.size())},
+                                       dtype_info);
+        shared_tensor_data = reinterpret_cast<uint8_t*>(sample[component_index].raw_mutable_data());
+      } else {
+        DALI_ENFORCE(size % static_cast<int64_t>(dtype_info.size()) == 0,
+                     "Index file at " + configs_[current_wds_shard_index_] +
+                         " reporting component sizes different to actual");
+        sample[component_index].ShareData(
+            shared_tensor_data, size, {size / static_cast<int64_t>(dtype_info.size())}, dtype_info);
+      }
+      sample_was_set[component_index] = true;
+    }
+  }
+  return shared_tensor_data;
+}
+
+inline void WebdatasetLoader::MarkCached(std::vector<Tensor<CPUBackend>>& sample,
+                                         std::vector<char>& sample_was_set,
+                                         const std::string& extension,
+                                         const std::string& source_info) const {
+  DALIMeta meta;
+  meta.SetSourceInfo(source_info);
+  meta.SetSkipSample(true);
+
+  for (size_t component_index : ext_map_.at(extension)) {
+    if (!sample_was_set[component_index]) {
+      sample[component_index].Reset();
+      sample[component_index].SetMeta(meta);
+      sample[component_index].set_type(TypeTable::GetTypeInfo(dtypes_[component_index]));
+      sample[component_index].Resize({0});
+      sample_was_set[component_index] = true;
+    }
+  }
+}
+
+void WebdatasetLoader::ReadSample(vector<Tensor<CPUBackend>>& sample) {
+  MoveToNextShard(GetCurrentSampleIndex());
+  auto& current_wds_shard = wds_shards_[current_wds_shard_index_];
+  auto& current_sample = wds_shards_metadata_[current_wds_shard_index_][current_sample_index_];
+  current_wds_shard.SeekArchive(current_sample.start_offset);
+
+  vector<char> sample_was_set(sample.size(), false);
```

Comment:
Why `char`? That's exactly a case for `vector<bool>` or our `bitmask` - or even `SmallVector`.
```suggestion
  vector<char> sample_was_set(sample.size(), false);
```